In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
road_accident = pd.read_csv('uk_road_accident.csv')

In [3]:
road_accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
road_accident['Accident Date'] = road_accident['Accident Date'].astype('str')
road_accident['Accident Date'] = road_accident['Accident Date'].str.strip()
road_accident['Accident Date'] = road_accident['Accident Date'].str.replace('/', '-')

In [5]:
road_accident['Accident Date'] = pd.to_datetime(road_accident['Accident Date'], dayfirst=True, errors='coerce')

In [6]:
road_accident['Year'] = road_accident['Accident Date'].dt.year
road_accident['Month_Number'] = road_accident['Accident Date'].dt.month
road_accident['Month'] = road_accident['Accident Date'].dt.month_name()
road_accident['Day'] = road_accident['Accident Date'].dt.day
road_accident['DayofWeek'] = road_accident['Accident Date'].dt.dayofweek  # Monday=0, Sunday=6

<h1>CHECK FOR NULL VALUES</h1>

In [7]:
road_accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
Year                           0
Month_Number                   0
Month                          0
Day                            0
DayofWeek                      0
dtype: int64

In [8]:
road_accident['Latitude'].mean()

np.float64(52.553865761110956)

In [9]:
road_accident['Latitude'] = road_accident['Latitude'].fillna(road_accident['Latitude'].mean())

In [10]:
road_accident['Longitude'].mean()

np.float64(-1.431210368502073)

In [11]:
road_accident['Longitude'] = road_accident['Longitude'].fillna(road_accident['Longitude'].mean())

In [12]:
road_accident['Road_Surface_Conditions'] = road_accident['Road_Surface_Conditions'].fillna('unaccounted')
road_accident['Road_Type'] = road_accident['Road_Type'].fillna('unaccounted')

In [13]:
road_accident['Urban_or_Rural_Area'].mode()

0    Urban
Name: Urban_or_Rural_Area, dtype: object

In [14]:
road_accident['Urban_or_Rural_Area'] = road_accident['Urban_or_Rural_Area'].fillna(road_accident['Urban_or_Rural_Area'].mode()[0])

road_accident['Weather_Conditions'] = road_accident['Weather_Conditions'].fillna('unaccounted')

In [15]:
road_accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month_Number               0
Month                      0
Day                        0
DayofWeek                  0
dtype: int64

<H1>CHECKING THE DATA TYPE</H1>

In [16]:
road_accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 19 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  object        
 1   Accident_Severity        660679 non-null  object        
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  object        
 5   District Area            660679 non-null  object        
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  object        
 10  Road_Type                660679 non-null  object        
 11  Urban_or_Rural_Area      660679 non-null  object        
 12  Weather_Conditio

In [17]:
road_accident['Index'] = road_accident['Index'].astype('category') 
road_accident['Accident_Severity'] = road_accident['Accident_Severity'].astype('category')
road_accident['Light_Conditions'] = road_accident['Light_Conditions'].astype('category')
road_accident['District Area'] = road_accident['District Area'].astype('category')
road_accident['Road_Surface_Conditions'] = road_accident['Road_Surface_Conditions'].astype('category') 
road_accident['Road_Type'] = road_accident['Road_Type'].astype('category') 
road_accident['Urban_or_Rural_Area'] = road_accident['Urban_or_Rural_Area'].astype('category') 
road_accident['Weather_Conditions'] = road_accident['Weather_Conditions'].astype('category') 
road_accident['Vehicle_Type'] = road_accident['Vehicle_Type'].astype('category')
road_accident['Month'] = road_accident['Month'].astype('category')
road_accident['Year'] = road_accident['Year'].astype('category')

In [18]:
road_accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 19 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  category      
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            660679 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  category      
 5   District Area            660679 non-null  category      
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  category      
 10  Road_Type                660679 non-null  category      
 11  Urban_or_Rural_Area      660679 non-null  category      
 12  Weather_Conditio

<H1>QUESTIONS AND 25 INSIGHT</H1>

<h2>What day of the week have the most and least accidents?</h2>

In [19]:
road_accident['DayofWeek'].value_counts() # Monday = 0, Sunday = 6

DayofWeek
5    107178
2     99558
3     99511
4     97900
1     94550
6     89302
0     72680
Name: count, dtype: int64

<h3>1. The data shows that saturday = 5 is the most prone day to accidents </h3>

<h3>2. The data shows that Monday = 0 is the least prone day to accidents</h3>

<h2>What year have the most and least number of accidents?</h2>

In [20]:
road_accident['Year'].value_counts()

Year
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<h3>3. The data shows that 2019 has the most number of accidents</h3>

<h3>4. The data shows that 2022 has the least number of accidents</h3>

<h2>Which district area has the most and least number of casualties?</h2>

In [21]:
road_accident.groupby('District Area')['Number_of_Casualties'].sum().idxmax()

'Birmingham'

In [22]:
road_accident.groupby('District Area')['Number_of_Casualties'].sum().idxmin()

'Clackmannanshire'

<h3>5. The data shows that the district area with most number of casualties is Birmingham</h3>

<h3>6. The data shows that the district area with least number of casualties is Clackmannanshire</h3>

<h2>What weather condition does accidents least and most recorded?</h2>

In [23]:
road_accident['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
unaccounted               14128
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h3>7. The data shows that accidents mostly occured in Fine and no high winds weather condition</h3>

<h3>8. The data shows that accidents least occured in Snowing + high winds weather condition</h3>

<h2>What road type has the least and most number of accidents?</h2>

In [24]:
road_accident.groupby('Road_Type').size()

Road_Type
Dual carriageway       99424
One way street         13559
Roundabout             43992
Single carriageway    492143
Slip road               7041
unaccounted             4520
dtype: int64

<h3>9. The data shows that Single carriageway type of road has the most number of accidents </h3>

<h3>10. The data shows that Slip road type of road has the least number of accidents if the unaccounted is exempted
 </h3>

<h2>What type of area does accidents happen the most?</h2>

In [25]:
road_accident.groupby('Urban_or_Rural_Area').size()

Urban_or_Rural_Area
Rural          238990
Unallocated        11
Urban          421678
dtype: int64

<h3>11. The result shows that accidents mostly happens on urban areas</h3>

<h2>What month does accidents least and mostly happen?</h2>

In [26]:
road_accident.groupby('Month').size()

Month
April        51744
August       53913
December     51836
February     49491
January      52872
July         57445
June         56481
March        54086
May          56352
November     60424
October      59580
September    56455
dtype: int64

<h3>12. According to the result, the month of november has the most numbers of accidents recorded</h3>

<h3>13. According to the result, the month of february has the least numbers of accidents recorded</h3>

<h2>The distribution of accidents by accident severity and vehicle type</h2>

In [27]:
road_accident.groupby(['Accident_Severity','Vehicle_Type']).size().unstack()

Vehicle_Type       Agricultural vehicle  Bus or coach (17 or more pass seats)  \
Accident_Severity                                                               
Fatal                                21                                   325   
Serious                             282                                  3373   
Slight                             1644                                 22180   

Vehicle_Type          Car  Data missing or out of range  \
Accident_Severity                                         
Fatal                6577                             0   
Serious             66461                             0   
Slight             424954                             6   

Vehicle_Type       Goods 7.5 tonnes mgw and over  \
Accident_Severity                                  
Fatal                                        216   
Serious                                     2321   
Slight                                     14770   

Vehicle_Type       Goods over 3.5t. and under 7.5t  \
Accident_Severity                                    
Fatal                                           67   
Serious                                        857   
Slight                                        5172   

Vehicle_Type       Minibus (8 - 16 passenger seats)  \
Accident_Severity                                     
Fatal                                            29   
Serious                                         276   
Slight                                         1671   

Vehicle_Type       Motorcycle 125cc and under  Motorcycle 50cc and under  \
Accident_Severity                                                          
Fatal                                     189                         95   
Serious                                  2031                       1014   
Slight                                  13049                       6494   

Vehicle_Type       Motorcycle over 125cc and up to 500cc  \
Accident_Severity                                          
Fatal                                                105   
Serious                                             1014   
Slight                                              6537   

Vehicle_Type       Motorcycle over 500cc  Other vehicle  Pedal cycle  \
Accident_Severity                                                      
Fatal                                339             70            6   
Serious                             3457            767           39   
Slight                             21861           4800          152   

Vehicle_Type       Ridden horse  Taxi/Private hire car  \
Accident_Severity                                        
Fatal                         0                    155   
Serious                       0                   1771   
Slight                        4                  11368   

Vehicle_Type       Van / Goods 3.5 tonnes mgw or under  
Accident_Severity                                       
Fatal                                              467  
Serious                                           4554  
Slight                                           29139

<h3>14. According to the result, The vehicle that has the most number in 3 types of accident severity is the Car</h3>

<h3>15. According to the result, The vehicle that has the least number in 3 types of accident severity is the Ridden Horse</h3>

<h2>What road conditions has the most and least numbers of accidents?</h2>

In [28]:
road_accident['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     447821
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
unaccounted                726
Name: count, dtype: int64

<h3>16. According to the result, the road condition with the most numbers of accidents is dry road</h3>

<h3>17. According to the result, the road condition with the least numbers of accidents is Flood over 3cm. deep road</h3>

<h2>What light condition has the most and least numbers of accident? </h2>

In [29]:
road_accident['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h3>18. According to the result, daylight has the most number of accidents recorded</h3>

<h3>19. According to the result, Darkness - lights unlit has the least number of accidents recorded</h3>

<h2>What is the maximum number of casualties in a single accident?</h2>

In [30]:
road_accident['Number_of_Casualties'].value_counts()

Number_of_Casualties
1     504041
2     107971
3      30776
4      11156
5       4123
6       1525
7        539
8        214
9        112
10        58
11        36
13        24
12        21
14        13
15        11
17         8
19         7
16         7
18         6
22         5
21         4
29         3
40         2
28         2
24         2
27         2
42         2
25         1
32         1
68         1
45         1
62         1
47         1
48         1
26         1
43         1
Name: count, dtype: int64

<h3>20. According to the result, the maximum casualties of accidents ever recorded is 504041</h3>

<h2>What is the number of accidents per month in each year?</h2>

In [31]:
accidentpermonth = road_accident.groupby(['Month', 'Year']).size()
accidentpermonth.unstack()

Year        2019   2020   2021   2022
Month                                
April      14125  13394  12715  11510
August     15044  13366  13415  12088
December   14708  13794  13709   9625
February   13253  14353  10950  10935
January    15355  14133  13417   9967
July       15862  14630  14300  12653
June       15528  14205  13936  12812
March      15049  13494  13202  12341
May        15833  14336  13811  12372
November   16559  14770  15473  13622
October    15528  15684  14834  13534
September  15271  14432  13792  12960

<h3>21. Based on the result, November is consistenly the peak month of accidents across all four years with two consequtive above 9% rate (2021, and 2022). Hoevwer, October also has high percentage with three ratings of 9% and above but November slightly beats it in most years.</h3>

<h3>22. Based on the result, February has the lowest percentage of accidents especially 2021 at just 6.7%. It suggests that road accidents tend to be more common in October, while February is generally a safer month.</h3>

<h2>Where do the latitude and longitude values fall within the data sets?</h2>

In [32]:
road_accident[['Latitude', 'Longitude']].describe()

Latitude      Longitude
count  660679.000000  660679.000000
mean       52.553866      -1.431210
std         1.406896       1.383302
min        49.914430      -7.516225
25%        51.490696      -2.332191
50%        52.315755      -1.411761
75%        53.453444      -0.232901
max        60.757544       1.762010

<h3>23. The result shows that the spans of the data set cover both northern/sothern and eastern/western regions</h3>

<h2>How many vehicles are usually involve in accidents?</h2>

In [33]:
road_accident['Number_of_Vehicles'].describe()

count    660679.000000
mean          1.831255
std           0.715269
min           1.000000
25%           1.000000
50%           2.000000
75%           2.000000
max          32.000000
Name: Number_of_Vehicles, dtype: float64

<h3>24. According to the result, Usually 2 vehicles are involve around</h3>

<h2>Is there a correlation between longitude and number of casualties?</h2>

In [34]:
road_accident['Longitude'].corr(road_accident['Number_of_Casualties'])

np.float64(-0.0404056457884545)

<h3>25. The result shows that there is no correlation between longitude and number of casualties.</h3>